clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2209,17 +2218,67 @@ namespace {
         DumpExpression(llvm::outs(), E);
     }
 
-    void VisitUnaryOperator(UnaryOperator *E, CheckedScopeSpecifier CSS) {
-      if (E->getOpcode() == UO_AddrOf)
-        S.CheckAddressTakenMembers(E);
+    BoundsExpr *CheckUnaryOperator(UnaryOperator *E, CheckedScopeSpecifier CSS, SideEffects SE) {
+      UnaryOperatorKind Op = E->getOpcode();
+      Expr *SubExpr = E->getSubExpr();
 
-      if (!E->isIncrementDecrementOp())
-        return;
+      // Recursively compute rvalue bounds for the subexpression,
+      // without performing bounds checking.  Since TraverseStmt still checks
+      // all substatements, enabling bounds checking here could cause
+      // duplicate side effects for an expression.  In future refactoring
+      // stages, this call to RValueBounds will be replaced with a call to
+      // TraverseStmt with side effects enabled (once TraverseStmt no longer
+      // always checks substatements).
+      BoundsExpr *SubExprBounds = RValueBounds(SubExpr, CSS, 
+                                               SideEffects::Disabled);
 
-      bool NeedsBoundsCheck = AddBoundsCheck(E->getSubExpr(), OperationKind::Other, CSS);
-      if (NeedsBoundsCheck && DumpBounds)
-          DumpExpression(llvm::outs(), E);
-      return;
+      // Perform checking with side effects, if enabled.
+      if (SE == SideEffects::Enabled) {
+        if (Op == UO_AddrOf)
+          S.CheckAddressTakenMembers(E);
+
+        if (E->isIncrementDecrementOp()) {
+          bool NeedsBoundsCheck = AddBoundsCheck(SubExpr, 
+                                                 OperationKind::Other, CSS);
+          if (NeedsBoundsCheck && DumpBounds)
+            DumpExpression(llvm::outs(), E);
+        }
+      }
+
+      // `*e` is not an rvalue.
+      // TraverseStmt (and CheckUnaryOperator) may be called on non-rvalues,
+      // so this is not unexpected behavior.
+      if (Op == UnaryOperatorKind::UO_Deref)
+        return CreateBoundsInferenceError();
```

Comment:
Since CheckUnaryOperator combines bounds inference and checking, and is called on every unary operator visited by TraverseStmt, is no longer guaranteed or expected that this (rvalue) bounds inference will only be done on rvalue unary operators. Adding in the previous unreachable call causes many test failures in Checked C, including static_checking/bounds_decl_checking.c